test/basic.test.js

```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
   await Promise.all(clients.map(c => c.end()))
 })
 
-test('Subscribed clients receive updates', async function (t) {
-  t.plan(10, 'each client should receive a message')
+test('Subscribed clients receive updates - QoS 1', async function (t) {
+  t.plan(20, 'each client should receive a message')
   t.tearDown(helper.closeBroker)
 
   await helper.startBroker()
 
-  var msg = {
-    topic: 'subscribers/topic',
-    payload: 'Hello world',
-    qos: 1,
-    retain: false
+  async function testQos (qos) {
+    var msg = {
+      topic: 'subscribers/topic',
+      payload: 'Hello world',
+      qos: qos,
+      retain: false
+    }
+
+    var subscribers = []
+    var received = 0
+
+    for (let i = 0; i < 10; i++) {
+      subscribers.push(helper.startClient())
+    }
+
+    subscribers = await Promise.all(subscribers)
+
+    function onMessage (topic, message) {
+      if (topic === msg.topic) {
+        if (received++ < 10 || qos === 2) {
+          t.pass('message received')
+        }
+      }
+    }
+
+    for (let i = 0; i < subscribers.length; i++) {
+      const client = subscribers[i]
+      await client.subscribe(msg.topic)
+      client.once('message', onMessage)
+    }
+
+    var publisher = await helper.startClient()
+
+    await publisher.publish(msg.topic, msg.payload, msg)
+    await helper.delay(100)
+
+    await Promise.all(subscribers.map(c => c.end()))
+    await publisher.end()
   }
 
-  var subscribers = []
+  await testQos(1)
+  await testQos(2)
+})
+
+test('Connect clean=false', async function (t) {
+  t.plan(1)
+  t.tearDown(helper.closeBroker)
+
+  const options = { clientId: 'pippo', clean: false }
+
+  await helper.startBroker()
+
+  var publisher = await helper.startClient('mqtt', options)
+
+  await publisher.subscribe('my/topic')
+
+  await publisher.end(true)
+
+  publisher = await helper.startClient('mqtt', options)
+
+  publisher.on('message', function (topic, message) {
+    if (topic === 'my/topic') {
+      t.pass('Subscription has been restored')
+    }
+  })
+
+  await publisher.publish('my/topic', 'I\'m alive', { qos: 1 })
+
+  await helper.delay(200)
```

Comment:
Is this needed? timeouts make tests brittle

Reply:
Unfortunally, I don't know why but I haven't find a way to make tap waiting, it always autoends and without this it calls tearDown before the message is received

Reply:
Would you like me to replace this using faketimers?